# How do I handle missing values in pandas?

In [ ]:
import pandas as pd

In [4]:
ufo=pd.read_csv('http://bit.ly/uforeports')

In [5]:
ufo.tail()

City Colors Reported Shape Reported State              Time
18236   Grant Park             NaN       TRIANGLE    IL  12/31/2000 23:00
18237  Spirit Lake             NaN           DISK    IA  12/31/2000 23:00
18238  Eagle River             NaN            NaN    WI  12/31/2000 23:45
18239  Eagle River             RED          LIGHT    WI  12/31/2000 23:45
18240         Ybor             NaN           OVAL    FL  12/31/2000 23:59

In the above output for row 18236 under Colors Reported column we see value as NaN rather than any color which indicates missing value and stands for 'Not a Number'. Below are few methods to deal with null values.

###### 16.1 - isnull( )

In [6]:
ufo.isnull().tail()

City  Colors Reported  Shape Reported  State   Time
18236  False             True           False  False  False
18237  False             True           False  False  False
18238  False             True            True  False  False
18239  False            False           False  False  False
18240  False             True           False  False  False

the way isnull works is if the value is missing or isnull then it will return True else False.

###### 16.2 - notnull( )

In [8]:
ufo.notnull().tail()

City  Colors Reported  Shape Reported  State  Time
18236  True            False            True   True  True
18237  True            False            True   True  True
18238  True            False           False   True  True
18239  True             True            True   True  True
18240  True            False            True   True  True

notnull works opposite of isnull() here it returns true if the value is not null

In [9]:
ufo.isnull().sum()

City                  25
Colors Reported    15359
Shape Reported      2644
State                  0
Time                   0
dtype: int64

above view shows number of missing values in each columns.

Below is the example on how it works

In [10]:
pd.Series([True,False,True]).sum()

2

Dummy series contains Boolean values i.e 2 True and 1 False . While perfroming mathmetical operations on them pandas converts True to 1 and False to 0

For ufo.isnull().sum() output is at column level . Refrencing to previous notes by default sum is doing axis 0 i.e at combing rows for each column . This can be checked in reverse manner below is the example.

In [14]:
ufo.isnull().sum(axis=1).head()

0    1
1    1
2    1
3    1
4    1
dtype: int64

In [15]:
ufo.head()

City Colors Reported Shape Reported State             Time
0                Ithaca             NaN       TRIANGLE    NY   6/1/1930 22:00
1           Willingboro             NaN          OTHER    NJ  6/30/1930 20:00
2               Holyoke             NaN           OVAL    CO  2/15/1931 14:00
3               Abilene             NaN           DISK    KS   6/1/1931 13:00
4  New York Worlds Fair             NaN          LIGHT    NY  4/18/1933 19:00

above we can see the for each row one value is null .

In [17]:
ufo[ufo.City.isnull()]

City Colors Reported Shape Reported State              Time
21     NaN             NaN            NaN    LA    8/15/1943 0:00
22     NaN             NaN          LIGHT    LA    8/15/1943 0:00
204    NaN             NaN           DISK    CA   7/15/1952 12:30
241    NaN            BLUE           DISK    MT    7/4/1953 14:00
613    NaN             NaN           DISK    NV    7/1/1960 12:00
1877   NaN          YELLOW         CIRCLE    AZ    8/15/1969 1:00
2013   NaN             NaN            NaN    NH     8/1/1970 9:30
2546   NaN             NaN       FIREBALL    OH  10/25/1973 23:30
3123   NaN             RED       TRIANGLE    WV  11/25/1975 23:00
4736   NaN             NaN         SPHERE    CA   6/23/1982 23:00
5269   NaN             NaN            NaN    AZ   6/30/1985 21:30
6735   NaN             NaN      FORMATION    TX     4/1/1992 2:00
7208   NaN             NaN         CIRCLE    MI   10/4/1993 17:30
8828   NaN             NaN       TRIANGLE    WA  10/30/1995 21:30
8967   NaN             NaN        VARIOUS    CA   12/8/1995 18:00
9273   NaN             NaN       TRIANGLE    OH     5/1/1996 3:00
9388   NaN             NaN           OVAL    CA   6/12/1996 12:00
9587   NaN             NaN            EGG    FL   8/24/1996 15:00
10399  NaN             NaN       TRIANGLE    IL   6/15/1997 23:00
11625  NaN             NaN         CIRCLE    TX     6/7/1998 7:00
12441  NaN             RED       FIREBALL    WA  10/26/1998 17:58
15767  NaN             NaN      RECTANGLE    NV   1/21/2000 11:30
15812  NaN             NaN          LIGHT    NV     2/2/2000 3:00
16054  NaN           GREEN            NaN    FL    3/11/2000 3:30
16608  NaN             NaN         SPHERE    NY   6/15/2000 15:00

Above code is used to filter DataFrame where city column is null. This returns 25 rows which is basically raw data for ufo.isnull().sum()

#### Options to deal with Null

###### Drop Nulls

In [18]:
ufo.shape

(18241, 5)

In [19]:
ufo.dropna(how='any').shape

(2486, 5)

dropna drops rows if value for any column is missing.By default dropna follows how=any . 

In [21]:
ufo.dropna().shape

(2486, 5)

In [22]:
ufo.dropna(how='all').shape

(18241, 5)

(how='all') drops if all the values are missing.

In [25]:
ufo.dropna(subset=['City','Shape Reported'],how = 'any').shape

(15576, 5)

Implies that if any of those two('City','Shape Reported') is missing for a given row then drop it

###### Filling Missing Values

In [26]:
ufo['Shape Reported'].value_counts(dropna=False)

LIGHT        2803
NaN          2644
DISK         2122
TRIANGLE     1889
OTHER        1402
CIRCLE       1365
SPHERE       1054
FIREBALL     1039
OVAL          845
CIGAR         617
FORMATION     434
VARIOUS       333
RECTANGLE     303
CYLINDER      294
CHEVRON       248
DIAMOND       234
EGG           197
FLASH         188
TEARDROP      119
CONE           60
CROSS          36
DELTA           7
CRESCENT        2
ROUND           2
PYRAMID         1
DOME            1
FLARE           1
HEXAGON         1
Name: Shape Reported, dtype: int64

we can see that NaN has a count of 2644. missing value can be updated with 'NotReceived' Below is te process.

In [27]:
ufo['Shape Reported'].fillna(value='NotReceived',inplace=True)

In [28]:
ufo['Shape Reported'].value_counts(dropna=False)

LIGHT          2803
NotReceived    2644
DISK           2122
TRIANGLE       1889
OTHER          1402
CIRCLE         1365
SPHERE         1054
FIREBALL       1039
OVAL            845
CIGAR           617
FORMATION       434
VARIOUS         333
RECTANGLE       303
CYLINDER        294
CHEVRON         248
DIAMOND         234
EGG             197
FLASH           188
TEARDROP        119
CONE             60
CROSS            36
DELTA             7
CRESCENT          2
ROUND             2
PYRAMID           1
HEXAGON           1
DOME              1
FLARE             1
Name: Shape Reported, dtype: int64

NaN has been replaced by NotReceived    2644